[자동차 경주 구현] 밀러 미션 제출합니다.

oungsi2000 opened the pull request:
리뷰 잘부탁드립니다

개인적으로 궁금한 점
1. Configure 클래스를 .env 나 .properties 파일로 저장하는게 나은지, 혹은 enum으로 저장하는 것이 낫다고 생각하시나요?
2. 현재는 모든 public 메서드를 테스트하지 않고 기능 명세의 핵심 로직만 (Service 클래스의 public 메서드) 테스트하고 있습니다, 번거롭더라도 견고한 프로그램을 작성하기 위해서 모든 메서드의 테스트를 작성하는 것이 나을까요?
3. 현재는 InputValidatoService 에서 검증 실패 시 illegalArgumentExcpetion을 던집니다. 컨트롤러에서 (Application.kt)에서 예외 핸들러를 만들어 모든 예외를 컨트롤러에서 따로 처리하도록 두는게 낫다고 생각하시나요?

medAndro committed:
docs: 구현 기능 목록 작성

medAndro committed:
feat: 프로젝트 기본 구조 작성

- 출력할 문자열들을 관리하는 Messages enum 추가
- 자동차 이름과 경주 횟수를 입력받는 로직 추가

medAndro committed:
feat: 입력값에 대한 예외 처리 구현

- InputValidator 클레스에 자동차 이름과 시도 횟수 예외를 테스트하는 메서드 구현

oungsi2000 committed:
feat(Car.kt) : Car 객체 구현

oungsi2000 committed:
feat:전체 기능 완성

oungsi2000 committed:
fix:랜덤값 버그 수정

oungsi2000 committed:
refactor:전체 리팩토링

oungsi2000 committed:
feat:테스트 코드 작성

medAndro committed:
refactor: 코드 컨벤션 및 테스트코드 호환성 수정

- klint 컨벤션 적용
- 테스트코드에서 줄 끝 문자의 윈도우/맥 호환을 위한 확장함수 추가

medAndro committed:
feat: 문자열 예외 테스트 케이스 추가

oungsi2000 committed:
merge test

medAndro committed:
refactor: MVC 구조로 변경

-  Application.kt 내 서비스 로직을 inputValidatorService으로 분리

oungsi2000 committed:
docs:readme 작성

oungsi2000 committed:
docs:readme 수정

lee-ji-hoon commented on README.md:
```diff
@@ -1 +1,29 @@
\ No newline at end of file
+# kotlin-racingcar
+
+### 1. 자동차 이름 및 시도 횟수 초기화
+   - 사용자로부터 자동차 이름과 시도 횟수를 입력받아 게임을 초기화합니다.
+   - 자동차 이름은 쉼표(,)로 구분하여 입력받고, 각 자동차 이름은 1자 이상 5자 이하, 중복 없이 영어, 한글, 숫자, 공백으로 구성되어야 합니다.
+   - 시도 횟수는 1 이상의 양의 정수 (8자리 이하)로 입력받습니다.
+   - 입력 과정에서 유효성 검사를 수행하고, 잘못된 입력에 대해서는 적절한 예외 처리를 진행합니다.
+   - 입력받은 자동차 이름과 시도 횟수를 기반으로 Car 객체를 생성하고, 게임을 준비합니다.
```
구현할 기능 목록과 커밋이 완벽하게 일치는 안하는거 같은데 한 번 체크포인트를 만들어서 관리 해보면 어떨까요?

기능 목록들을 기반으로 체크포인트를 만들어서 관리를 하다 보면 전체적인 흐름을 시작 전에 정리 할 수 있기 때문이죠! 물론 FUNCTION_LIST도 잘 만들어주셨지만 전체적인 구현 단위를 한 번 정해보는 연습도 해보면 좋을거 같네요!

oungsi2000 replied:
미리 FUNCTION_LIST를 토대로 기능을 구현하고 readme는 gpt로 보기 편하게 돌려서 그렇습니다. 말씀주신대로 다음에는 readme의 기능 리스트들을 기준으로 커밋을 하도록 하겠습니다

lee-ji-hoon commented on src/main/kotlin/racingcar/InputValidatorService.kt:
```diff
@@ -0,0 +1,56 @@
+import racingcar.domain.Messages
+
+class InputValidatorService {
+    fun getValidatedCarNames(input: String): List<Car> {
```
`InputValidatorService` 의 역할은 Input의 Validate역할이라고 보여지는데요.

그 역할 관점으로 봤을 때 `List<Car>` 을 반환하는게 맞을까요? 아니면 Validate인 Car을 Name만을 넘기는게 맞을까요?

oungsi2000 replied:
Car의 Name만을 넘기는게 맞다고 생각합니다. 당시에는 내부적으로 이미 List를 생성하고 있기 때문에 그냥 List<Car> 를 넘기는 것이 더 효율적이라고 생각했던 것 같습니다. 만약 객체를 생성하는 로직이 매우 비싸고 검증하는 로직 내부에서 무조건 객체를 생성해야 하는 경우엔 어떤 식으로 처리하는 것이 좋을까요?

lee-ji-hoon commented on src/main/kotlin/racingcar/InputValidatorService.kt:
```diff
@@ -0,0 +1,56 @@
+class InputValidatorService {
+    fun getValidatedCarNames(input: String): List<Car> {
+        emptyCheck(input)
+        val names = input.split(",").map { it.trim() }
```
trim까지 신경써주셨다니 너무 좋습니다 👍

lee-ji-hoon commented on src/main/kotlin/racingcar/InputValidatorService.kt:
```diff
@@ -0,0 +1,56 @@
+
+    private fun emptyCheck(input: String) {
+        require(input.isNotBlank()) { Messages.ERROR_EMPTY_INPUT.message }
+    }
```
개인적으로 이렇게까지 함수화를 할 필요가 있을까 싶습니다.

```kotlin
fun getValidatedRaceCount(input: String): Int {
    emptyCheck(input)
    val number = positiveInputCheck(input)
    tryCountLengthCheck(number)
    return number
}

fun getValidatedRaceCount(input: String): Int {
    require(input.isNotBlank()) { Messages.ERROR_EMPTY_INPUT.message }
    val number = positiveInputCheck(input)
    tryCountLengthCheck(number)
    return number
}
```

`emptyCheck` 내부를 들어가기 전까지는 `require` 인지 아닌지 알 수가 없기 때문이죠. 

그렇기에 누군가 이 함수를 사용할 때 `getValidatedRaceCount` 만 보고 exception이 나는 것을 모르고 사용하게 될 수도 있습니다.

lee-ji-hoon replied:
그게 아니라면 `require` 을 prefix로 해버리는 것도 좋은 방법 같네요!

oungsi2000 replied:
확실히 그럴수 있겠네요 ! 반영하겠습니다

lee-ji-hoon commented on src/main/kotlin/racingcar/InputValidatorService.kt:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private fun availableNameCheck(name: String) {
+        val regex = Regex("^[가-힣a-zA-Z0-9]*$")
```
Regex 까지 활용하시다니 좋습니다!

`Regex의 경우 생성 비용이 비싸다.` 라는 것이 보편적이며, 관련해서는 `이펙티브 코틀린 아이템 45. 불필요한 객체 생성을 피하라` 를 읽어보시는 것도 좋습니다!

oungsi2000 replied:
require(name.all { it.isLetterOrDigit() || it.isWhitespace() }) 로 Regex를 대체하였습니다.

lee-ji-hoon commented on src/main/kotlin/racingcar/InputValidatorService.kt:
```diff
@@ -0,0 +1,56 @@
+    private fun positiveInputCheck(input: String): Int {
+        val number =
+            runCatching { input.toInt() }
+                .getOrElse { throw IllegalArgumentException(Messages.ERROR_NOT_POSITIVE.message) }
```
```suggestion
val number = input.toIntOrNull() ?: throw IllegalArgumentException(Messages.ERROR_NOT_POSITIVE.message)
```

이런 형태도 가능해보이는데요! kotlin에는 다양한 확장함수들이 존재하다보니 한 번씩 둘러보고 사용해보시는 것을 추천드립니다 👍

lee-ji-hoon commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,10 @@
+
+class Car(
+    val name: String,
+    var distance: Int = 0,
```
name은 val인데 distance은 var로 선언이 되어있네요. 

var로 선언한 이유는 아마 변경이 자주 일어나기 때문일거 같아요. 그런데 var로 했을 때 문제는 없을까요?

oungsi2000 replied:
외부에서 임의의 값을 직접 설정할 수 있는 문제가 있습니다. 

```
var distance: Int = 0
        private set
```

로 var로 선언하여 내부적으로 distance의 값을 변경 가능케 하되, setter 로직을 날려주어 해당 문제를 해결했습니다

lee-ji-hoon commented on src/main/kotlin/racingcar/domain/Configure.kt:
```diff
@@ -0,0 +1,7 @@
+package racingcar.domain
+
+class Configure {
+    companion object {
+        const val RANDOM_SEED: Int = 123
+    }
+}
```
> Configure 클래스를 .env 나 .properties 파일로 저장하는게 나은지, 혹은 enum으로 저장하는 것이 낫다고 생각하시나요?

1. .env 나 properties를 고민하신 이유
2. enum 과 현재 companion object 그리고 그냥 object 각 어떤 차이점들이 있을까요?

이것들을 다시 고민해보면 무엇이 더 어울릴지 정답이 나올거 같네요!

oungsi2000 replied:
enum의 장점 중 하나는 컴파일 시 타입 안정성을 확보할 수 있는 것입니다. 하지만 상수 값의 경우에는 여러 타입이 들어올 수 있어(Int, String 등), .env 와 .properties에 저장하는 것이 더 자연스럽다고 생각합니다.

하지만 이러한 설정 파일들에 저장하는 경우, 값을 읽어올 때, 무조건 String으로 변환된다는 단점이 있습니다. 따라서, 값과 타입을 동시에 저장이 가능한 코드의 형태인 object로 다시 구현하였습니다. 이러한 상수 값은 메모리 효율을 위해 싱글톤이 적합하다고 생각하여 object 키워드를 사용하였습니다

lee-ji-hoon replied:
음 그러면 그냥 Object 클래스여도 됐지 않았을까? 라는 생각이 있네요

굳이 동반 객체로 만들 필요가 있었을까요?

lee-ji-hoon commented on src/test/kotlin/ApplicationTest.kt:
```diff
@@ -0,0 +1,156 @@
+import java.io.PrintStream
+import kotlin.random.Random
+
+class ApplicationTest {
```
`ApplicationTest` 의 이름만 봤을때 어떤 테스트들이 존재할지 감이 오실까요? 저는 Application에 대한 전반적인 생명주기라던지 Application이 생성이 될 때 초기화가 잘되는지 등의 테스트가 존재할거 같은데요.

그런데 실제로는 Validator를 테스트한다던지 작은 단위가 테스트되고 있는데요.
그러면 작은 단위로 Unit Test를 나눠보는 연습을 해봅시다!

oungsi2000 replied:
확실히 테스트 또한 작은 단위로 나눠야 할 것 같습니다 ! 반영하겠습니다

lee-ji-hoon commented on src/main/kotlin/racingcar/InputValidatorService.kt:
```diff
@@ -0,0 +1,56 @@
+
+    private fun duplicateNameCheck(names: List<String>) {
+        require(names.size == names.distinct().size) { Messages.ERROR_DUPLICATE_NAME.message }
+    }
```
> 현재는 InputValidatoService 에서 검증 실패 시 illegalArgumentExcpetion을 던집니다. 컨트롤러에서 (Application.kt)에서 예외 핸들러를 만들어 모든 예외를 컨트롤러에서 따로 처리하도록 두는게 낫다고 생각하시나요?

Application에서 예외 핸들러를 만들어 모든 예외를 컨트롤러에서 따로 처리한다는 말씀이 제가 잘 이해가 안되는데 간단한 예시를 알 수 있을까요? 

지금도 Application에서 처리를 안할 뿐 `InputValidatorService` 에서는 이미 `illegalArgumentExcpetion`를 전파하고 있는데요. 여기서 `예외 핸들러를 만들어서 처리를 한다.` 라는게 각 `Check` 함수들에서 validation이 false이면 따로 Custom한 Exception을 만들어서 전파를 하고 그것을 처리한다는걸까요?

oungsi2000 replied:
GlobalExceptionHandler를 생각해주시면 좋을 것 같습니다. 

현재는 InputValidatorService에서 예외가 발생할 수 있는 경우가 있고, try-catch로 예외를 잡지 않고 있습니다. 
제 말은 Controller에서 Service를 호출할 때, Service에서 발생한 예외를 Controller로 올려(예외 번역을 하여) Controller 단에서 모든 예외를 처리하는 것입니다.

예시 :
Controller에서 View, Service 호출 -> View에서 예외 발생 -> View에서 try-catch 하지 않고 Controller로 예외 전파 -> Controller에서 전파된 예외 처리

이렇게 처리하는 것입니다. 

말씀주신 상황에서는 validation이 false이면 현재는 validator에서 예외를 발생하고 프로그램이 바로 종료되지만, validator를 호출하는 view로 예외 전파 ->view에서 Controller로 예외를 전파하여 최종적으로는 모든 예외를 Controller에서 관리하는것이 좋을까? 라는 뜻이었습니다.

다시 코니의 리뷰를 보고 리팩터링 하였을 때, 든 생각은 '지금은 필요 없다' 였습니다. 현재 애플리케이션에서는 잘못된 인자를 받을 경우, 프로그램이 바로 뻗게 되어 있어, 의미가 없다고 생각하였습니다. 오히려 코드가 과도하게 복잡해질 수 있다고 생각했습니다. 

만약 잘못된 인자를 받았을 경우 다른 비즈니스 로직을 수행하도록 프로그램이 설계되었을 경우, 그리고 그런 로직이 꽤나 많을 경우, 한꺼번에 여러 예외를 관리할 수 있으니 도입을 하는 것이 좋다고 생각합니다.

lee-ji-hoon replied:
네 좋습니다.

저는 `GlobalExceptionHandler`가 필요한 상황은 충분히 존재하지만 `GlobalExceptionHandler`가 정말로 필요한지에 대해서는 많은 고민이 필요하다고 생각을해요. `GlobalExceptionHandler` 을 함으로써 모든 레이어의 에러를 상위에서 알아야 하거나, 불필요하게 잡는 상황(앱이 터져야하는데 빈화면만 보인다던지)을 초래할 수 있기 때문이죠.

> 만약 잘못된 인자를 받았을 경우 다른 비즈니스 로직을 수행하도록 프로그램이 설계되었을 경우, 그리고 그런 로직이 꽤나 많을 경우, 한꺼번에 여러 예외를 관리할 수 있으니 도입을 하는 것이 좋다고 생각합니다.

그런데 이 측면에서는 특정 상황에서는 좋기도 합니다. `에러가 생겼을 때 로그를 보낸다.` 이런 것들이 필요하다고 할 때 매번 만들때마다 이것을 만들어야 한다는 것을 놓치기도 쉽기 때문이죠.

---

지금은 우선 밀러의 생각이 틀리지 않았다고 생각을 하니 이대로 계속 구현을 해가보면 좋을거 같네요!

lee-ji-hoon commented on src/main/kotlin/racingcar/RaceService.kt:
```diff
@@ -0,0 +1,34 @@
+    fun race(
+        raceCount: Int,
+        cars: List<Car>,
+    ): String {
```
Race에서 Race에 필요한 정보들은 잘 받아주신거 같습니다.

그런데 race 라는 함수를 호출한뒤 String을 반환했을 때 이 String이 무엇인지 바로 알 수 있을까요?

oungsi2000 replied:
확실히 String을 바로 반환하는 경우, 사용자가 String이 무엇인지 알기도 힘들고, 또 View의 역할을 Service에서 수행하고 있는 것 같다는 생각이 들었습니다. gameResult 일급 컬렉션을 만들어 해당 컬렉션을 반환하도록 하여 게임 결과를 반환한다는 의미를 조금 명확히 하도록 반영했습니다

lee-ji-hoon replied:
아무래도 로또에서 일급 컬렉션을 이미 공부를 해서 정답을 찾았군요 👍

lee-ji-hoon commented on src/test/kotlin/ApplicationTest.kt:
```diff
@@ -0,0 +1,156 @@
+
+    @Test
+    @DisplayName("Int 자료형으로 입력 가능한 1이상의 자연수만 허용한다")
+    fun t3() {
+        setInput("a\n0")
+        assertThrows<IllegalArgumentException> { Application().run() }
+    }
```
이 테스트가 Application에 관련한 테스트일까요? `IllegalArgumentException` 이 발생한다는게 메인 주제 같은데 그럼 어떤 테스틀 해봐야할지 알 수 있을거 같습니다.

그러면 자연스럽게 `setInput` 이 필요할지도 알게 될거 같네요!

oungsi2000 replied:
이 부분도 전체 ApplicationTest가 아닌 validator의 기능이어서, ValidatorTest를 따로 만들었습니다 !

lee-ji-hoon commented on src/test/kotlin/ApplicationTest.kt:
```diff
@@ -0,0 +1,156 @@
+        )
+    }
+
+    @Test
+    @DisplayName("랜덤 값 추출 테스트, 확률이 60%~60.01% 사이여야 한다")
+    fun t6() {
+        val random = Random(RANDOM_SEED)
+        var count = 0
+        val osName = System.getProperty("os.name")
+        assertThat(osName).contains("MAc")
+        repeat(100000000) { if (random.nextInt(0, 10) >= 4) count++ }
+        assertThat(count)
+            .isGreaterThan(60000000)
+            .isLessThan(60010000)
+    }
```
이 테스트가 Application에 맞는 테스트 인지 그리고 정말 필요한 테스트인지 고민이 필요해보이네요.

`60~60.01%` 인지를 알아서 어떤 이점이 있을까요?

oungsi2000 replied:
처음 제이슨 코치가 강의를 하셨을 때 라이브러리 테스트를 하시는 것을 보고 영감을 받아 작성한 코드였습니다. 하지만 다시 보니 코니의 말씀대로 불필요해 보이네요...

lee-ji-hoon commented on src/main/kotlin/racingcar/Application.kt:
```diff
@@ -0,0 +1,38 @@
+}
+
+class Application {
+    private val inputValidatorService = InputValidatorService()
+    private val raceService = RaceService()
+    private val cars = mutableListOf<Car>()
```
`private val cars = mutableListOf<Car>()` 가 진짜 필드 변수로 있어야 할 이유가 있을지 고민해보면 좋을거 같습니다. 필드 변수의 경우 해당 클래스가 살아있을 때까지는 계속 유지가 되기도 하며, 모든 함수에서 접근이 가능하기 때문이죠

oungsi2000 replied:
코드를 리팩토링 하다보니 자연스레 사라지게 되었습니다...다만 필드 변수를 없애고 메서드의 인자로 cars를 받게끔 한다면 각 메서드의 호출이 복잡해지고, 만약 Application 객체의 메서드 1의 인자의 cars와 메서드2의 인자가 다를 경우, 오히려 더 좋지 않은 상황이 나올 것 같습니다. 필드를 줄이는 것이 코틀린 프로그래밍 스타일이지만 저는 해당 상황의 경우에는 필드로 유지하는 것도 좋다고 생각합니다.

lee-ji-hoon replied:
`필드를 줄이는 것이 코틀린 프로그래밍 스타일` 여기서 코틀린의 프로그래밍 스타일인지는 저는 그리 느끼지는 않은거 같아요. 

1. 생성자로 받을 때
2. 필드에 선언할 때
3. 함수에 선언할 때

보편적으로는 이렇게 존재할텐데 각 필요에 따라서 선언을 하기 때문인데요.

생성자로 받아아하는 상황에서는 이 객체가 생성될 때 필요한 객체인 것이고 필드는 이 객체 내부에서 전역적으로 사용이 될 때 그리고 마지막으로 함수는 해당 함수 스코프에서만 살아있으면 될 때 사용을 해야 한다 생각을 하네요.

lee-ji-hoon reviewed (changes requested):
안녕하세요 밀러님! 자동차 경주 구현 리뷰어 코니입니다!

README 되게 꼼꼼히 작성해주신 덕분에 저도 해당 내용을 보면서 확인을 많이 했네요!

그런데 저희가 지금 생각해야 하는건 객체의 역할, 그리고 테스트 입니다.

그런 관점에서 봤을 때 현재 객체의 역할이 뚜렷한지, 테스트가 꼼꼼하게 전부 이뤄졌는지를 확인해보면 좋을거 같네요.

---

> 현재는 모든 public 메서드를 테스트하지 않고 기능 명세의 핵심 로직만 (Service 클래스의 public 메서드) 테스트하고 있습니다, 번거롭더라도 견고한 프로그램을 작성하기 위해서 모든 메서드의 테스트를 작성하는 것이 나을까요?

기능 명세의 핵심 로직만 테스트 하는 것이 베스트가 맞습니다. 하지만 현재 테스트 구현하신 것은 핵심 로직을 전부 구현하지는 못하신 상태인데요. 그런 관점에 서 봤을 때 우선 public 메서드들 전부 구현한다 생각하고 구현 후 테스트와 도메인적인 것들이 익숙해지면 그때가서 핵심 로직들만 빼서 테스트 해보는건 어떨까요?

medAndro committed:
refactor: 자동차의 전진 여부를 로직 수정

- 전진 여부를 자동차 객체 내부에서 자동차에게 주어지는 정수값에 의해 결정하도록 변경
- 값의 범위를 넘어가면 예외가 발생하는 코드 추가

medAndro committed:
refactor: OutputView 도입 및 stringBuilder 제거

medAndro committed:
refactor: 자동차 이름을 자동차 객체 내부에서 검사하도록 변경

medAndro committed:
refactor: Car 객체 생성 위치와 컨트롤러 코드 수정

- Car객체를 InputValidator가 아닌 RaceService에서 생성하도록 변경
- 사용자와의 입출력을 콘솔로 받는 View 사용
- run()을 통해 게임 시작시 게임의 초기화를 진행

medAndro committed:
refactor: 테스트 유연성을 위한 의존성 주입 도입

medAndro committed:
fix: 자동차 이름 검증 로직 수정

- 이름 중간에 공백이 존재하는 경우를 허용하도록 수정

medAndro committed:
test: InputValidator클래스의 테스트 코드 추가

medAndro committed:
test: 자동차 모델의 테스트 코드 추가

medAndro committed:
refactor: 테스트 용이성 향상을 위한 인터페이스 적용

medAndro committed:
test: RaceService 클래스의 테스트 코드 추가

medAndro committed:
refactor: 테스트 전용 인터페이스 분리

lee-ji-hoon reviewed (approved):
안녕하세요 밀러 개인 사정으로 인해 첫 미션 진행이 어렵다고 전달을 받았습니다~!

그래도 그 다음이 리팩토링이니까 지금 현재 미션은 머즈하고 시간이 되실 때 다시 요청 주시면 리뷰 해드리겠습니다!
